### PR TITLE
Add test and benchmark for explicit dot GEMM

### PR DESCRIPTION
**Pull request by brunomazzottiamd**
### Summary

This PR adds missing test and benchmark features to the explicit dot GEMM Triton kernel developed in the scope of https://github.com/ROCm/triton-internal/issues/169. It also adds GEMM implemented with `tl.dot` to the mix so we can easily compare it to PyTorch GEMM and Triton GEMM implemented with explicit dot, both in terms of correctness and performance.

### Help / How to run

```bash
python python/perf-kernels/multreduce_matmul_kernel.py -h
```

```
usage: multreduce_matmul_kernel.py [-h] [-M M] [-N N] [-K K] [--use-bias] [--use-dot] {run,bench}

C = A * B + BIAS matrix multiplication kernel for small matrices (M ≤ 8)

positional arguments:
  {run,bench}  mode of operation:
                 run: run Triton kernel for a given (M, N, K) shape
                 bench: benchmark performance for target shapes

options:
  -h, --help   show this help message and exit

kernel shape arguments:
  -M M         rows of matrix A (must be less or equal to 8)
  -N N         columns of matrix A / rows of matrix B
  -K K         columns of matrix B
  --use-bias   use BIAS vector
  --use-dot    use tl.dot for dot product
```

### Running the Triton kernel for a single shape

```bash
python python/perf-kernels/multreduce_matmul_kernel.py run -M 1 -N 8192 -K 28672
```

### Checking correctness of a single target shape

```bash
pytest -vvv python/perf-kernels/multreduce_matmul_kernel.py::test_matmul[1-4096-4096-True]
```

### Checking correctness of all target shapes

```bash
pytest -vvv python/perf-kernels/multreduce_matmul_kernel.py
```

### Benchmarking all target shapes

```bash
python python/perf-kernels/multreduce_matmul_kernel.py bench
```

Sample benchmark result on MI300X:
```
fp16_multreduce_matmul_kernel:
     M        N        K  Torch (GiB/s)  Triton Dot (GiB/s)  Triton Multreduce (GiB/s)
0  1.0   8192.0  28672.0        3861.00             2508.09                    2305.36
1  1.0   6144.0   6144.0        2034.46             2640.22                    1714.58
2  1.0   4096.0   4096.0        2048.43             1191.51                    1107.05
3  2.0  16384.0  16384.0        2418.02             2794.15                    2715.15
4  1.0   4096.0   3078.0        1066.43              839.09                     369.35
5  1.0     23.0     31.0           0.21                0.05                       0.05
6  1.0     23.0    128.0           0.84                0.20                       0.20
```

### CODE REVIEW

File: python/perf-kernels/multreduce_matmul_kernel.py [hunk 1]
```diff
@@ -1,45 +1,340 @@
...
 
-# Kernel that implements GEMM with explicit multiply-reduce instructions for small block sizes.
-# Based on **tune_gemm** `matmul_kernel` from commit `cf44637` (see `triton-mlir` branch).
+
+def get_triton_autotune_configs() -> list[triton.Config]:
+    # yapf: disable
+    return [
+        triton.Config(
+            {
+                "BLOCK_SIZE_M": 1, "BLOCK_SIZE_N": 64, "BLOCK_SIZE_K": 256, "waves_per_eu": 0,
```
Comment (rahulbatra85) [inline python/perf-kernels/multreduce_matmul_kernel.py:56]: Do you think any more config option will make sense here? Or will they add too much time to autotuning?

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:56]: Hi Rahul.

Let me explain my rationale behind these 3 autotuning configs. There's 4 target GEMM shapes, `tune_gemm.py` gave me 3 configs as the best ones, 2 target shapes share the same best config. So I added these 3 configs as autotuning options.

Probably we won't be choosing a good config if a shape other them the target ones is selected. How did you get your autotuning configs for RMSNorm and Softmax? I'm open to autotuning suggestion as well, feel free to suggest anything you think is appropriate. 

About autotuning time: To be honest, I didn't even worry about this. As far as I can see it's fast enough.

Thanks for your feedback.

Comment (rahulbatra85) [reply python/perf-kernels/multreduce_matmul_kernel.py:56]: @brunomazzottiamd I just came up by making good guesses. I varied the num_warps and waves_per_eu. I do have a lot of auto-configs and for one kernel(softmax) the run takes a few mintues(but not too long). Also, I was running mostly shapes with lower number of rows.

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:56]: Hi @rahulbatra85. I'm going to rebase this and add more configs, varying `num_warps`. Do you have a study suggestion on `waves_per_eu`? It's always zero in `tune_gemm.py`, I want to better understand the meaning of this parameter.

Comment (rahulbatra85) [reply python/perf-kernels/multreduce_matmul_kernel.py:56]: As I understand, it's a hint to the compiler to schedule waves_per_eu number of waves per execution unit(SIMD unit). More waves per eu will help with better utilization of GPU.

Comment (rahulbatra85) [reply python/perf-kernels/multreduce_matmul_kernel.py:56]: I am not sure why it's set to 0 in tune_gemm.py. May be check with Shucai.

By the way, this config option is only valid for AMD GPUs

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: I'm closing this discussion since it's very old and I added more configs to autotuning. I think any issue would be better covered in a new discussion.

File: python/perf-kernels/multreduce_matmul_kernel.py [hunk 2]
```diff
@@ -1,45 +1,464 @@
...
+def triton_matmul(triton_provider: str, a: Tensor, b: Tensor, bias: Optional[Tensor]) -> Tensor:
+    assert triton_provider in ["triton-dot", "triton-multreduce"]
+
+    M: int
+    N: int
+    K: int
+    M, K = a.shape
+    _, N = b.shape
+
+    c: Tensor = torch.empty((M, N), device=a.device, dtype=a.dtype)
```
Comment (xiaohuguo2023) [inline python/perf-kernels/multreduce_matmul_kernel.py:224]: this should be torch.zeros instead of torch.empty ?

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:224]: Thanks for your comment Xiaohu.

I think we can use either `torch.empty` or `torch.zeros` here because the Triton kernel initializes the accumulator as `accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)`.

However, I can use `torch.zeros` instead of `torch.empty` if you think it's better.

Comment (xiaohuguo2023) [reply python/perf-kernels/multreduce_matmul_kernel.py:224]: I think you will have correctness issue if your output is not initialised ?

Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: Resolved by https://github.com/ROCm/triton/pull/637/commits/02ddbfd90b6ba11b6f334a6a4f26d4dfa701a5b9.

**Review by vgokhale — APPROVED** [review]

File: python/perf-kernels/multreduce_matmul_kernel.py [hunk 3]
```diff
@@ -1,45 +1,464 @@
...
+# ------------------------------------------------------------------------
+# It's a pytest suite, you can run it with `pytest multreduce_matmul_kernel.py`.
+# You can also run a single test with
+# `multreduce_matmul_kernel.py::test_matmul[M-N-K-use_bias]`.
+
+
+def get_target_shapes() -> list[tuple[int, int, int]]:
+    # yapf: disable
+    return [
+        (1, 8192, 28672),
```
[the remark this answers is not in the document]
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:309]: No problem, let's do it! I think it's a good test to stress the algorithm.

[the remark this answers is not in the document]
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py:309]: @vgokhale, shape $(1, 23, 31)$ breaks the kernel, unit test fails. I'm going to investigate it.

[the remark this answers is not in the document]
Comment (brunomazzottiamd) [reply python/perf-kernels/multreduce_matmul_kernel.py]: The bug is fixed, but I don't know why the current code works and the previous code doesn't work.

**Review by vgokhale — APPROVED** [review]
Thank you!
